core/src/main/java/org/infinispan/persistence/sifs/Index.java

```diff
@@ -267,20 +267,27 @@ public CompletionStage<Object> handleRequest(IndexRequest indexRequest) {
       return indexRequest;
    }
 
-   public void deleteFileAsync(int fileId) {
+   public CompletionStage<Object> ensureRunOnLast(Runnable runnable) {
       AtomicInteger count = new AtomicInteger(flowableProcessors.length);
-      IndexRequest deleteFile = IndexRequest.syncRequest(() -> {
-         // After all indexes have ensured they have processed all requests - the last one will delete the file
-         // This guarantees that the index can't see an outdated value
+      IndexRequest request = IndexRequest.syncRequest(() -> {
          if (count.decrementAndGet() == 0) {
-            fileProvider.deleteFile(fileId);
-            log.tracef("Deleted file %s", fileId);
-            compactor.releaseStats(fileId);
+            runnable.run();
          }
       });
       for (FlowableProcessor<IndexRequest> flowableProcessor : flowableProcessors) {
-         flowableProcessor.onNext(deleteFile);
+         flowableProcessor.onNext(request);
       }
+
+      return request;
```

Comment:
Actually this doesn't work at all, since the request will be completed by the first node, not the last. I would say don't even have it return. And we probably could just remove the complete call in the `accept` method as no one relies upon its completion.